### PR TITLE
Added sqlite3 driver

### PR DESCRIPTION
**Pull request by ErenYeager713**

### CODE REVIEW

File: src/libs/common.go [hunk 1]
```diff
@@ -99,7 +99,7 @@ func SetDefaultConfig() {
 	viper.SetDefault("application.cluster.cluster-info-from", "k8sclient")
 
 	// Database config
-	viper.SetDefault("database.driver", "mysql")
+	viper.SetDefault("database.driver", "sqlite333")
```
Comment (nyrahul) [inline src/libs/common.go]: what is `sqlite333`?

File: src/libs/dbHandler.go [hunk 2]
```diff
@@ -26,6 +26,12 @@ func GetNetworkPolicies(cfg types.ConfigDB, cluster, namespace, status string) [
 			return results
 		}
 		results = docs
+	} else if cfg.DBDriver == "sqlite3" {
```
Comment (nyrahul) [inline src/libs/dbHandler.go:29]: Lets just use `sqlite` as the driver name.

File: src/libs/dbHandler_test.go [hunk 3]
```diff
@@ -20,7 +20,7 @@ const Unmet = "unmet expectation error: "
 // ==================== //
 
 func TestGetNetworkPolicies(t *testing.T) {
-	// prepare mock mysql
+	// prepare mock sqlite
```
Comment (nyrahul) [inline src/libs/dbHandler_test.go:103]: dont change existing testcase of mysql since we will support mysql as well. Add new test cases for sqlite.

File: src/networkpolicy/helperFunctions.go [hunk 4]
```diff
@@ -528,26 +528,27 @@ func updateDNSFlows(networkLogs []types.KnoxNetworkLog) {
...
 		if log.DNSRes != "" && log.DNSResIPs != nil {
 			domainName := log.DNSRes
+			newDNSIPs := log.DNSResIPs
 
 			// udpate DNS to IPs map
-			if ips, ok := DomainToIPs[domainName]; ok {
-				for _, ip := range ips {
-					if !libs.ContainsElement(ips, ip) {
-						ips = append(ips, ip)
+			if dnsIps, ok := DomainToIPs[domainName]; ok {
```
Comment (nyrahul) [inline src/networkpolicy/helperFunctions.go]: Why are these changes part of this PR??

File: src/plugin/cilium.go [hunk 5]
```diff
@@ -253,6 +253,97 @@ func ConvertCiliumFlowToKnoxNetworkLog(ciliumFlow *cilium.Flow) (types.KnoxNetwo
 	return log, true
 }
 
+func ConvertSQLiteCiliumLogsToKnoxNetworkLogs(docs []map[string]interface{}) []types.KnoxNetworkLog {
```
Comment (nyrahul) [inline src/plugin/cilium.go:256]: We do not use this function anywhere I believe. Can we remove this function?

File: src/plugin/kubearmor.go [hunk 6]
```diff
@@ -50,6 +50,61 @@ func ConvertKnoxSystemPolicyToKubeArmorPolicy(knoxPolicies []types.KnoxSystemPol
 	return results
 }
 
+func ConvertSQLiteKubeArmorLogsToKnoxSystemLogs(docs []map[string]interface{}) []types.KnoxSystemLog {
```
Comment (nyrahul) [inline src/plugin/kubearmor.go:53]: This function is not used as well from anywhere.

File: src/protobuf/v1/worker/worker.proto [hunk 7]
```diff
@@ -18,6 +18,7 @@ message WorkerRequest {
     string namespace = 4;
     string clustername = 5;
     string labels = 6;
+    string fromsource = 7;
```
Comment (nyrahul) [inline src/protobuf/v1/worker/worker.proto]: This change should not be part of this PR. Please rebase the code.

**Review by nyrahul — CHANGES_REQUESTED** [review]
Thanks. Mostly looks good, few comments in the PR.

File: src/conf/local-file.yaml [hunk 8]
```diff
@@ -26,7 +26,7 @@ application:
     cluster-mgmt-url: "http://localhost:8080"
 
 database:
-  driver: mysql
+  driver: sqlite3
```
Comment (seswarrajan) [inline src/conf/local-file.yaml:29]: @ErenYeager713 
Kindly revert the change back to mysql. Once after validation we can make this change on need basis. 
Also we need to decide on the persistent directory that we will be mounting the DB to in case of K8S.

File: src/conf/local.yaml [hunk 9]
```diff
@@ -24,7 +24,7 @@ application:
     cluster-mgmt-url: "http://localhost:8080"
 
 database:
-  driver: mysql
+  driver: sqlite3
```
Comment (seswarrajan) [inline src/conf/local.yaml:27]: @ErenYeager713
Kindly revert the change back to mysql. Once after validation we can make this change on need basis.
Also we need to decide on the persistent directory that we will be mounting the DB to in case of K8S.

File: src/config/configManager_test.go [hunk 10]
```diff
@@ -49,7 +49,7 @@ feed-consumer:
       buffer: 50
 
 database:
-  driver: mysql
+  driver: sqlite3
```
Comment (seswarrajan) [inline src/config/configManager_test.go:52]: @ErenYeager713
Kindly revert the change back to mysql. Once after validation we can make this change on need basis.
Also we need to decide on the persistent directory that we will be mounting the DB to in case of K8S.

File: src/libs/sqliteHandler.go [hunk 11]
```diff
@@ -0,0 +1,748 @@
...
+
+func connectSQLite(cfg types.ConfigDB) (db *sql.DB) {
+	if MockDB != nil {
+		return MockDB
+	}
+
+	dbconn := cfg.DBUser + ":" + cfg.DBPass + "@tcp(" + cfg.DBHost + ":" + cfg.DBPort + ")/" + cfg.DBName
+	fmt.Println(dbconn)
+	//db, err := sql.Open(cfg.DBDriver, dbconn)
+	db, err := sql.Open(cfg.DBDriver, "/home/vagrant/accuknox.db")
```
Comment (seswarrajan) [inline src/libs/sqliteHandler.go]: @ErenYeager713 Please do not hard-code any paths. Get this path from local-file config and use the same across the code.

File: src/libs/sqliteHandler.go [hunk 12]
```diff
@@ -0,0 +1,748 @@
...
+}
+
+func connectSQLite(cfg types.ConfigDB) (db *sql.DB) {
+	if MockDB != nil {
+		return MockDB
+	}
+
+	dbconn := cfg.DBUser + ":" + cfg.DBPass + "@tcp(" + cfg.DBHost + ":" + cfg.DBPort + ")/" + cfg.DBName
+	fmt.Println(dbconn)
+	//db, err := sql.Open(cfg.DBDriver, dbconn)
```
Comment (seswarrajan) [inline src/libs/sqliteHandler.go]: @ErenYeager713  Remove this commented line.

**Review by seswarrajan — CHANGES_REQUESTED** [review]
Please rebase the code and update the PR.
Also address the review comments

File: src/plugin/kubearmor.go (hunk 6 above)
Comment (ErenYeager713) [reply src/plugin/kubearmor.go:53]: This function is used here https://github.com/ErenYeager713/auto-policy-discovery/blob/dev/src/systempolicy/systemPolicy.go

File: src/protobuf/v1/worker/worker.proto (hunk 7 above)
Comment (ErenYeager713) [reply src/protobuf/v1/worker/worker.proto]: Okay

File: src/networkpolicy/helperFunctions.go (hunk 4 above)
Comment (ErenYeager713) [reply src/networkpolicy/helperFunctions.go]: WIll rebase the code again

File: src/conf/local-file.yaml (hunk 8 above)
Comment (ErenYeager713) [reply src/conf/local-file.yaml:29]: Sure

File: src/libs/sqliteHandler.go (hunk 11 above)
Comment (ErenYeager713) [reply src/libs/sqliteHandler.go]: Okay

File: src/libs/dbHandler.go (hunk 2 above)
Comment (ErenYeager713) [reply src/libs/dbHandler.go:29]: If we don't specify `sqlite3` then it doesn't identify the sqlite driver

Comment (seswarrajan) [reply src/libs/dbHandler.go:29]: The driver name should match the name as per the driver value in local-file config file. So sqlite should work.

File: src/plugin/cilium.go (hunk 5 above)
Comment (ErenYeager713) [reply src/plugin/cilium.go:256]: It is used in cilium.go file itself
```
func ConvertCiliumNetworkLogsToKnoxNetworkLogs(dbDriver string, docs []map[string]interface{}) []types.KnoxNetworkLog {
	if dbDriver == "mysql" {
		return ConvertMySQLCiliumLogsToKnoxNetworkLogs(docs)
	} else if dbDriver == "sqlite3" {
                return ConvertSQLiteCiliumLogsToKnoxNetworkLogs(docs)
        } else if dbDriver == "mongo" {
		return ConvertMongodCiliumLogsToKnoxNetworkLogs(docs)
	} else {
		return []types.KnoxNetworkLog{}
	}
}
```

File: src/libs/dbHandler.go (hunk 2 above)
Comment (ErenYeager713) [reply src/libs/dbHandler.go:29]: Yeah so I tried with `sqlite` and it doesn't recognise the driver, also in the local-file the name of the driver is `sqlite3` , because anything else and it doesn't recognise the driver

File: src/systempolicy/systemPolicy.go [hunk 13]
```diff
@@ -1279,7 +1279,7 @@ func GenFileSetForAllPodsInCluster(clusterName string, pods []types.Pod, settype
 		} else {
 			if !reflect.DeepEqual(mergedfs, out[wpfs]) {
 				log.Info().Msgf("updating wpfs db entry for wpfs=%+v", wpfs)
-				err = libs.UpdateWorkloadProcessFileSetMySQL(CfgDB, wpfs, mergedfs)
+				err = libs.UpdateWorkloadProcessFileSetSQLite(CfgDB, wpfs, mergedfs)
```
Comment (nyrahul) [inline src/systempolicy/systemPolicy.go:1282]: This will not work when the driver is set to mysql. Note that we also have to support mysql.
CC @seswarrajan

**Review by nyrahul — CHANGES_REQUESTED** [review]
Note that we are not deprecating mysql. We are adding one more DB driver in the form of sqlite. Both mysql and sqlite needs to work.
With the current changes, mysql will stop working.
